src/core/qgsmaplayer.h

```diff
@@ -73,13 +73,17 @@ class CORE_EXPORT QgsMapLayer : public QObject
      */
     QgsMapLayer( QgsMapLayer::LayerType type = VectorLayer, const QString &name = QString::null, const QString &source = QString::null );
 
-    virtual ~QgsMapLayer();
+    /** Copy constructor.
+     * @param mapLayer the map layer to copy
+     * @param deep to copy unique id if true
+     */
+    QgsMapLayer( const QgsMapLayer &mapLayer, bool deep = true );
```

Comment:
The docs also need much more detail here, explaining what happens with data providers, edit buffers, selections, etc.